src/passes/OptimizeInstructions.cpp

```diff
@@ -841,6 +757,54 @@ struct OptimizeInstructions
           }
         }
       }
+      if (auto* constTrue = select->ifTrue->dynCast<Const>()) {
+        if (auto* constFalse = select->ifFalse->dynCast<Const>()) {
+          if (select->type == Type::i32 || select->type == Type::i64) {
+            auto trueValue = constTrue->value.getInteger();
+            auto falseValue = constFalse->value.getInteger();
+            if ((trueValue == 1LL && falseValue == 0LL) ||
+                (trueValue == 0LL && falseValue == 1LL)) {
+              Builder builder(*getModule());
+              // canonicalize "expr ? 0 : 1" to "!expr ? 1 : 0"
+              if (trueValue == 0LL) {
+                select->condition = optimizeBoolean(
+                  builder.makeUnary(EqZInt32, select->condition));
+              }
+              Expression* condition = nullptr;
+              if (Properties::emitsBoolean(select->condition)) {
+                // !x ? 1 : 0   ==>   !x
+                // x <=> y ? 1 : 0   ==>   x <=> y
+                condition = select->condition;
+              } else {
+                // expr ? 1 : 0   ==>   !!expr
+                if (getPassOptions().shrinkLevel > 0) {
```

Comment:
I don't get this. why is it checking the shrink level? why not always emit the shorter form? that would be shorter and has no downside I can see.

Reply:
`eqz(eqz(x))` is smaller but not all engines could optimize this. For example some interpreters without preopt will evaluate it as is. But I agree perhaps it make sense leave this renormalization to finalize optimization pass as I proposed in #2922. For now just use double eqz